src/meds_etl/flat.py

```diff
@@ -300,12 +300,114 @@ def process_csv_file(
+    )
+
+    grouped_by_time = all_events.group_by("patient_id", "time").agg(measurements=measurement)
+    
+    event = pl.struct(
+        pl.col("time"),
+        pl.col("measurements"),
+    )
+
+    grouped_by_patient = grouped_by_time.group_by("patient_id").agg(events=event.sort_by(pl.col("time")))
+
+    # We now have our data in the final form, grouped_by_patient, but we have to do one final transformation
+    # We have to convert from polar's large_list to list because large_list is not supported by huggingface
 
+    # We do this conversion using the pyarrow library
+
+    # Save and load our data in order to convert to pyarrow library
+    converted = grouped_by_patient.collect().to_arrow()
+
+    # Now we need to reconstruct the schema
+    if len(metadata_columns) == 0:
+        metadata_schema = pa.null()
+    else:
+        metadata_schema = pa.struct([(column, pa.string()) for column in metadata_columns])
+
+    desired_schema = meds.patient_schema(metadata_schema)
+
+    # All the larg_lists are now converted to lists, so we are good to load with huggingface
+    casted = converted.cast(desired_schema)
+
+    pq.write_table(casted, os.path.join(data_dir, f"data_{shard_index}.parquet"))
+
+
+def process_shard_duckdb(args):
+    """Given a specific shard of patients, use DuckDB to group their events into MEDS timelines and save to disk."""
+    import duckdb
+    shard_index: int = args[0]
+    temp_dir: str = args[1]
+    data_dir: str = args[2]
+
+    logging.info("Processing shard ", shard_index)
+    shard_dir: str = os.path.join(temp_dir, str(shard_index))
+    path_to_output: str = os.path.join(data_dir, f"data_{shard_index}.parquet")
+
+    if len(os.listdir(shard_dir)) == 0:
+        return
+    
+    # TODO (@Michael) -- Instead of COPY directly to .parquet,
+    # should we convert to polars via `.pl()` then apply the casting to MEDS schema?
+    duckdb.sql(f"""
+        DROP TABLE IF EXISTS all_events_{shard_index};
+        CREATE TABLE all_events_{shard_index} AS SELECT * FROM '{shard_dir}/*.parquet';
+        COPY (
+            SELECT
+                patient_id,
+                list({{ 
+                    'time' : time, 
+                    'measurements' : measurements 
+                }}) AS events
+            FROM (
+                SELECT
+                    patient_id,
+                    time,
+                    list({{
+                        'code' : code, 
+                        'text_value' : text_value, 
+                        'numeric_value' : numeric_value, 
+                        'datetime_value' : datetime_value, 
+                        'metadata' : metadata 
+                    }}) AS measurements
+                FROM all_events_{shard_index}
+                GROUP BY patient_id, time
+                ORDER BY patient_id, time ASC
+            )
+            GROUP BY patient_id
+            ORDER BY patient_id
+        ) TO '{path_to_output}' (FORMAT 'parquet', CODEC 'zstd');
```

Comment:
Are we sure this is generating parquet files with the correct schema? We want some sort of assert and/or cast here if possible.

Reply:
Yeah good q -- I left a `TODO` to convert this back into polars via `.pl()` and then we can use the same `cast` to schema code as before. Will make that change and report back.

Reply:
We don't need to convert it back to polars? That seems like it would add a lot of overhead. I would just open the output file with pyarrow and do a schema assert check.